core/store/src/trie/mem/updating.rs

```diff
@@ -1,29 +1,67 @@
+//! Structures and logic for updating in-memory trie.
+//!
+//! DISCLAIMER: This is in process of rewriting to generic structures.
+//! See #12324.
+//! For now we keep the old types together with new ones to change logic
+//! incrementally (for example, `GenericNodeOrIndex` and `OldOrUpdatedNodeId`).
+//! New methods will be prefixed with `generic_` to distinguish them from the
+//! old ones. When the old methods are removed, the prefix will be dropped.
+
 use super::arena::{ArenaMemory, ArenaMut};
 use super::flexible_data::children::ChildrenView;
 use super::metrics::MEM_TRIE_NUM_NODES_CREATED_FROM_UPDATES;
 use super::node::{InputMemTrieNode, MemTrieNodeId, MemTrieNodeView};
-use crate::trie::{Children, MemTrieChanges, TrieRefcountDeltaMap, TRIE_COSTS};
+use crate::trie::insert_delete::NodesStorage;
+use crate::trie::{
+    Children, MemTrieChanges, NodeHandle, StorageHandle, TrieNode, TrieNodeWithSize,
+    TrieRefcountDeltaMap, ValueHandle, TRIE_COSTS,
+};
 use crate::{NibbleSlice, RawTrieNode, RawTrieNodeWithSize, TrieChanges};
+use near_primitives::errors::StorageError;
 use near_primitives::hash::{hash, CryptoHash};
 use near_primitives::state::FlatStateValue;
 use std::collections::HashMap;
 use std::sync::Arc;
 
+/// For updated nodes, the ID is simply the index into the array of updated nodes we keep.
+pub type GenericUpdatedNodeId = usize;
+
+pub type UpdatedMemTrieNodeId = usize;
+
 /// An old node means a node in the current in-memory trie. An updated node means a
 /// node we're going to store in the in-memory trie but have not constructed there yet.
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub enum OldOrUpdatedNodeId {
-    Old(MemTrieNodeId),
-    Updated(UpdatedMemTrieNodeId),
+pub enum GenericNodeOrIndex<GenericTrieNodePtr> {
+    Old(GenericTrieNodePtr),
+    Updated(GenericUpdatedNodeId),
 }
 
-/// For updated nodes, the ID is simply the index into the array of updated nodes we keep.
-pub type UpdatedMemTrieNodeId = usize;
+pub type OldOrUpdatedNodeId = GenericNodeOrIndex<MemTrieNodeId>;
 
-/// An updated node - a node that will eventually become an in-memory trie node.
+/// Trait for trie values to get their length.
+pub trait HasLength {
```

Comment:
nit: perhaps name this something a little more specific, like HasValueLength; HasLength is a very generic name when reading the trait bound in another file.